WKWebView alert window is truncating title

youennf opened the pull request:
#### d02c05b3862111d6ae974f65d3a8288dfb0acc24
<pre>
WKWebView alert window is truncating title
<a href="https://bugs.webkit.org/show_bug.cgi?id=254958">https://bugs.webkit.org/show_bug.cgi?id=254958</a>
rdar://106327229

Reviewed by Jean-Yves Avenard.

Introduce a routine to create an alert dialog which is not truncating very long titles.
Use it wherever alerts are created in UIProcess code.

Manually tested.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _presentLockdownMode]):
* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::alertForPermission):
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(WebKit::displayStorageAccessAlert):
* Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.h: Added.
* Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.mm: Added.
(WebKit::createUIAlertController):
* Source/WebKit/UIProcess/ios/WKPasswordView.mm:
(-[WKPasswordView showPasswordFailureAlert]):
* Source/WebKit/UIProcess/ios/WKWebGeolocationPolicyDeciderIOS.mm:
(-[WKWebGeolocationPolicyDecider _executeNextChallenge]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController _showPhishingAlert]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/262909@main">https://commits.webkit.org/262909@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/97fdc575bf7eb6a66dbd8d70194b7a7c927a6436

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3068 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4413 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3399 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3147 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3109 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2627 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3034 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3392 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2696 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4208 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/889 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2675 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2523 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2726 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3950 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3072 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2464 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2701 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2677 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/737 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2689 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2892 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.mm:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebKit {
+
+UIAlertController* createUIAlertController(NSString *title, NSString *message)
```
Nit: `UIAlertController *createUIAlertController(NSString *title, NSString *message)`.

cdumez replied:
Should return a RetainPtr too, otherwise, "create" in the name makes it tempting for someone to release the value returned by this function.

pxlcoder commented on Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.h:
```diff
@@ -0,0 +1,38 @@
+
+namespace WebKit {
+
+UIAlertController* createUIAlertController(NSString *title, NSString *message);
```
Nit: `UIAlertController *createUIAlertController(NSString *title, NSString *message)`.

youennf replied:
OK, will change to RetainPtr< UIAlertController > though

jyavenard commented on Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.h:
```diff
@@ -0,0 +1,38 @@
+
+#if PLATFORM(IOS_FAMILY)
+
+#import "UIKitSPI.h"
```
is this needed?

jyavenard commented on Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.mm:
```diff
@@ -0,0 +1,42 @@
+
+#import "config.h"
+#import "UIAlertControllerUtilities.h"
+
```
should import SPI here instead

pxlcoder reviewed (commented):
Can any of these titles come from web content or do we control all the titles?

cdumez commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -3143,7 +3143,7 @@ - (void)_presentLockdownMode
                 appDisplayName = [[NSBundle mainBundle] objectForInfoDictionaryKey:(__bridge NSString *)kCFBundleNameKey];
 
-            UIAlertController *alert = [UIAlertController alertControllerWithTitle:[NSString stringWithFormat:WEB_UI_NSSTRING(@"Lockdown Mode is Turned On For “%@“", "Lockdown Mode alert title"), appDisplayName] message:message.get() preferredStyle:UIAlertControllerStyleAlert];
+            UIAlertController *alert = WebKit::createUIAlertController([NSString stringWithFormat:WEB_UI_NSSTRING(@"Lockdown Mode is Turned On For “%@“", "Lockdown Mode alert title"), appDisplayName], message.get());
```
If something is called create(), it should return a RetainPtr. Otherwise it is unclear if we need to release the returned value or not.

youennf replied:
Will do

pxlcoder commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -38,7 +38,7 @@
 #import "RemoteScrollingCoordinatorProxyIOS.h"
 #import "ScrollingTreeScrollingNodeDelegateIOS.h"
 #import "TapHandlingResult.h"
-#import "UIKitSPI.h"
```
I imagine this is getting pulled in from another header, but there is other SPI usage in this file and I think we should keep the import explicit.

Same comment applies to the other removals (unless there's one where we aren't actually using any other SPI).

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -27,7 +27,7 @@
 
 #if PLATFORM(IOS)
-#import "UIKitSPI.h"
+#import "UIAlertControllerUtilities.h"
```
Maybe we just move this import down below, since the rest of the code is compiled out otherwise?

youennf committed:
WKWebView alert window is truncating title
https://bugs.webkit.org/show_bug.cgi?id=254958
rdar://106327229

Reviewed by Jean-Yves Avenard.

Introduce a routine to create an alert dialog which is not truncating very long titles.
Use it wherever alerts are created in UIProcess code.

Manually tested.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _presentLockdownMode]):
* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::alertForPermission):
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(WebKit::displayStorageAccessAlert):
* Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.h: Added.
* Source/WebKit/UIProcess/ios/UIAlertControllerUtilities.mm: Added.
(WebKit::createUIAlertController):
* Source/WebKit/UIProcess/ios/WKPasswordView.mm:
(-[WKPasswordView showPasswordFailureAlert]):
* Source/WebKit/UIProcess/ios/WKWebGeolocationPolicyDeciderIOS.mm:
(-[WKWebGeolocationPolicyDecider _executeNextChallenge]):
* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenViewController.mm:
(-[WKFullScreenViewController _showPhishingAlert]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/262909@main
